src/v/transform/io.h

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "model/record.h"
+#include "model/record_batch_reader.h"
+#include "model/transform.h"
+
+namespace transform {
+namespace detail {
+
+/**
+ * A factory for sources and sinks.
+ *
+ * Use sink::factory or source::factory instead.
+ */
+template<typename T>
+class factory {
+public:
+    factory() = default;
+    factory(const factory&) = delete;
+    factory& operator=(const factory&) = delete;
+    factory(factory&&) = delete;
+    factory& operator=(factory&&) = delete;
+    virtual ~factory() = default;
+
+    virtual std::optional<std::unique_ptr<T>> create(model::ntp) = 0;
+};
+} // namespace detail
+
+/**
+ * The output sink for Wasm transforms.
+ */
+class sink {
+public:
+    sink() = default;
+    sink(const sink&) = delete;
+    sink& operator=(const sink&) = delete;
+    sink(sink&&) = delete;
+    sink& operator=(sink&&) = delete;
+    virtual ~sink() = default;
+
+    virtual ss::future<> write(ss::chunked_fifo<model::record_batch>) = 0;
```

Comment:
should this take a record_batch_reader?

Reply:
Then the driving of the transform is in the lower layers like the sink implementation or even possibly as we're in serde writing to the wire... Is that a common pattern we use?

In general I'm pro wrapping readers with a decorator pattern.

Reply:
the common pattern is that whatever the originating source is drives the pipeline through readers/consumers. this isn't a strict thing, tho. for example, raft unwraps data off the wire into an in-memory set of batches and passes that to storage. it still rewraps in a reader though because other users of storage actually do some streaming.

anyway, i think it's ok to leave as-is and see how things shake out with optimizations and improvements.